internal/services/v1/experimental.go

```diff
@@ -520,11 +533,30 @@ func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.B
 		})
 	}
 
+	// Run the checks in parallel.
 	if err := tr.StartAndWait(); err != nil {
 		return nil, es.rewriteError(ctx, err)
 	}
 
-	return resp, nil
+	// Compute hashes for each of the pairs.
+	pairHashes := map[*v1.BulkCheckPermissionPair]string{}
+	for _, pair := range pairs {
+		pairItemHash, err := computeBulkCheckPermissionItemHash(pair.Request)
+		if err != nil {
+			return nil, es.rewriteError(ctx, err)
+		}
+
+		pairHashes[pair] = pairItemHash
+	}
+
+	// Resort the results based on the incoming request order to make it easier to consume. This is not strictly
+	// necessary as each result contains the request that made it, but for the non-streaming version
+	// of this API, it makes sense.
+	slices.SortStableFunc(pairs, func(first *v1.BulkCheckPermissionPair, second *v1.BulkCheckPermissionPair) int {
+		return itemIndexByHash[pairHashes[first]] - itemIndexByHash[pairHashes[second]]
+	})
```

Comment:
I assume this is going to be an O(n log n) sort, right? You could do the sorting in linear time by re-using your map from above.

```
sortedPairs := make([]*v1.BulkCheckPermissionPair, 0, len(req.Items))
for _, pair := range pairs {
    currHash := computeBulkCheckPermissionItemHash(pair.Request)
    pairIdx := itemIndexByHash[currHash]
    sortedPairs[pairIdx] = pair
}
```

Reply:
Or if you don't want the added memory of another `sortedPairs` list, you could compute the hash and insert into `pairs` at the proper index in the first loop.

Reply:
Updated. Thanks for the suggestion